Dynamically switching from scrollbar-width:none to scrollbar-width:auto/thin doesn't bring back scrollbar

nmoucht opened the pull request:
#### 6115517674045ec83c36c570bf353e30718d5499
<pre>
Dynamically switching from scrollbar-width:none to scrollbar-width:auto/thin doesn&apos;t bring back scrollbar
<a href="https://bugs.webkit.org/show_bug.cgi?id=282419">https://bugs.webkit.org/show_bug.cgi?id=282419</a>
<a href="https://rdar.apple.com/139040042">rdar://139040042</a>

Reviewed by Simon Fraser.

Previously, scrollbar-width:none was implemented by nulling out the host layer on the
NSScrollerImp. When switching from scrollbar-width:none to scrollbar-width:auto/thin,
we wouldn&apos;t keep around the old scrollbar layer, which would cause the scrollbar to not
appear. We fix this by implementing scrollbar-width:none as nulling out the NSScrollerImp
when hidden by style, and keeping around the previous scrollbar layer on m_hostLayer to
be re-attached when switching back to scrollbar-width:auto/thin.

* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::setHiddenByStyle):
(WebCore::ScrollerMac::updateScrollbarStyle):
(WebCore::ScrollerMac::updatePairScrollerImps):
(WebCore::ScrollerMac::updateMinimumKnobLength):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):

Canonical link: <a href="https://commits.webkit.org/286057@main">https://commits.webkit.org/286057@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99e2aff9b92b87a600e5854da38fdfd7b653ac81

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54114 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27501 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79102 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25922 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76801 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63250 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1899 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58665 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16950 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77751 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48813 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64184 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39061 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45949 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21681 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24255 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67232 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80596 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2002 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1176 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66926 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2150 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64202 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66217 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10160 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8311 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11525 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1967 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4754 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1995 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/2915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2002 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
Could we make a copy of the paint WPT in `fast/scrolling/mac/scrollbars` and add:

` <!-- webkit-test-runner [ MockScrollbarsEnabled=false AsyncOverflowScrollingEnabled=true ] -->`

and `internals.setUsesOverlayScrollbars(false);` to emulate the wpt.fyi config?

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -428,7 +443,7 @@ - (void)invalidate
 void ScrollerMac::updateMinimumKnobLength(int minimumKnobLength)
 {
-    if (m_minimumKnobLength == minimumKnobLength)
+    if (m_minimumKnobLength == minimumKnobLength && minimumKnobLength)
```
Why the check for `minimumKnobLength` here? Explain in the commit message.

nmoucht replied:
Removed.

smfr reviewed (commented):
This needs a test or some new results or expectation changes.

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-width-dynamic-none-to-auto-expected.html:
```diff
@@ -0,0 +1,24 @@
+<!-- webkit-test-runner [ MockScrollbarsEnabled=false AsyncOverflowScrollingEnabled=true ] -->
+<!doctype html>
```
The normal form is
```
<!DOCTYPE html> <!-- webkit-test-runner....-->
```

nmoucht replied:
Fixed.

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-width-dynamic-none-to-auto.html:
```diff
@@ -0,0 +1,30 @@
+        testRunner.waitUntilDone();
+    internals.setUsesOverlayScrollbars(false);
+    requestAnimationFrame(() => requestAnimationFrame(() => {
+        document.getElementById('box').style.scrollbarWidth = 'auto';
```
The way I like to do this is `document.body.classList.add('changed')` and write a `body.changed .box` selector with the new style.

nmoucht replied:
Fixed.

nmoucht committed:
Dynamically switching from scrollbar-width:none to scrollbar-width:auto/thin doesn't bring back scrollbar
https://bugs.webkit.org/show_bug.cgi?id=282419
rdar://139040042

Reviewed by Simon Fraser.

Previously, scrollbar-width:none was implemented by nulling out the host layer on the
NSScrollerImp. When switching from scrollbar-width:none to scrollbar-width:auto/thin,
we wouldn't keep around the old scrollbar layer, which would cause the scrollbar to not
appear. We fix this by implementing scrollbar-width:none as nulling out the NSScrollerImp
when hidden by style, and keeping around the previous scrollbar layer on m_hostLayer to
be re-attached when switching back to scrollbar-width:auto/thin.

* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::setHiddenByStyle):
(WebCore::ScrollerMac::updateScrollbarStyle):
(WebCore::ScrollerMac::updatePairScrollerImps):
(WebCore::ScrollerMac::updateMinimumKnobLength):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):

Canonical link: https://commits.webkit.org/286057@main
